core/src/main/scala/sbt/librarymanagement/VersionNumber.scala

```diff
@@ -5,159 +5,188 @@ final class VersionNumber private[sbt] (
+     *
+     * Rule 10 means that build metadata is never considered for compatibility
+     *         we'll enforce this immediately by dropping them from both versions
+     * Rule 2 we enforce with custom extractors.
+     * Rule 4 we enforce by matching x = 0 & fully equals checking the two versions
+     * Rule 6, 7 & 8 means version compatibility is determined by comparing the two X values
+     * Rule 9 means pre-release versions are fully equals checked
+     */
     def isCompatible(v1: VersionNumber, v2: VersionNumber): Boolean =
-      doIsCompat(v1, v2) || doIsCompat(v2, v1)
+      doIsCompat(dropBuildMetadata(v1), dropBuildMetadata(v2))
+
     private[this] def doIsCompat(v1: VersionNumber, v2: VersionNumber): Boolean =
       (v1, v2) match {
-        case (v1, v2)
-            if (v1.size >= 2) && (v2.size >= 2) => // A normal version number MUST take the form X.Y.Z
-          (
-            v1._1.get,
-            v1._2.get,
-            v1._3.getOrElse(0L),
-            v1.tags,
-            v2._1.get,
-            v2._2.get,
-            v2._3.getOrElse(0L),
-            v2.tags
-          ) match {
-            case (0L, _, _, _, 0L, _, _, _) =>
-              // Major version zero (0.y.z) is for initial development. Anything may change at any time. The public API should not be considered stable.
-              equalsIgnoreExtra(v1, v2)
-            case (_, 0, 0, ts1, _, 0, 0, ts2) if ts1.nonEmpty || ts2.nonEmpty =>
-              // A pre-release version MAY be denoted by appending a hyphen and a series of dot separated identifiers
-              equalsIgnoreExtra(v1, v2)
-            case (x1, _, _, _, x2, _, _, _) =>
-              // Patch version Z (x.y.Z | x > 0) MUST be incremented if only backwards compatible bug fixes are introduced.
-              // Minor version Y (x.Y.z | x > 0) MUST be incremented if new, backwards compatible functionality is introduced
-              x1 == x2
-            case _ => equalsIgnoreExtra(v1, v2)
-          }
-        case _ => false
+        case (NormalVersion(0, _, _), NormalVersion(0, _, _))         => v1 == v2 // R4
+        case (NormalVersion(x1, _, _), NormalVersion(x2, _, _))       => x1 == x2 // R6, R7 & R8
+        case (PrereleaseVersion(_, _, _), PrereleaseVersion(_, _, _)) => v1 == v2 // R9
+        case _                                                        => false
       }
-    // Build metadata SHOULD be ignored when determining version precedence.
-    private[this] def equalsIgnoreExtra(v1: VersionNumber, v2: VersionNumber): Boolean =
-      (v1.numbers == v2.numbers) && (v1.tags == v2.tags)
+
+    // SemVer Spec Rule 10 (above)
+    private[VersionNumber] def dropBuildMetadata(v: VersionNumber) =
+      if (v.extras.isEmpty) v else VersionNumber(v.numbers, v.tags, Nil)
+
+    // SemVer Spec Rule 9 (above)
+    private[VersionNumber] def isPrerelease(v: VersionNumber): Boolean = v.tags.nonEmpty
```

Comment:
This doesn't capture the subtlety of Rule 9.

> Pre-release versions have a lower precedence than the associated normal version.

Here are some examples:

- 1.0.0-M1 is a pre-release of 1.0.0, and thus have lower precedence than 1.0.0.
- 1.1.0-M1 is a pre-release of 1.1.0, and thus have lower precedence than 1.1.0. However, since this is intended to be a pre-release of 1.1.0 series, a reasonable expectation is to assume compatibility with 1.0.0 series.

Thus my original code treats `case (_, 0, 0, ts1, _, 0, 0, ts2) if ts1.nonEmpty || ts2.nonEmpty` as a special case.

Reply:
> However, since this is intended to be a pre-release of 1.1.0 series, a reasonable expectation is to assume compatibility with 1.0.0 series.

SemVer states:

Reply:
Right, so in case of 1.0.0-M1 whose associated normal version is 1.0.0, the "might not" actually is meaningful. But this doesn't mean that every time we issue an M1, it's completely incompatible with sbt 1.x for example.

Reply:
I'll push a commit with the diff of allowing these to be compatible, so we can compare.